### ISSUE TITLE
Conflicts with existing old-style OpenCMISS installations

### ISSUE THREAD

**Issue by rondiplomatico** [OpenCMISS/manage#17]
**Problem**
If the old OpenCMISS system is installed on a computer, the USE_SYSTEM_XXX defaults cause trouble. As the OpenCMISSExtras etc are locatable via environment variables (PATH) that cmake also takes into account, it will pick up e.g. LAPACK or LIBXML2 from the old opencmissextras locations.
if so, the linking stage for consuming entities like hypre or libcellml is highly likely to fail. e.g. zlib is build locally, libxml2 is found under the old opencmissextras and libcellml tries to build binaries which link against a static libxml2.a that in turn knows nothing (without the old installation context) of zlib. boom.
**Suggested solution**
We need to provide a simple check to see if an old opencmiss installation is present and then switch to OC_COMPONENTS_SYSTEM "NONE" instead of "DEFAULT". this way no old library gets in the way.

any further thoughts?

**Comment by chrispbradley** [comment]
If the suggested solution is easy then sure. If not then I wouldn't be too stressed as the instructions for the new build can simply state to delete the old OpenCMISSExtras. The versions of dependencies are likely to be too old for the new code anyway.

**Comment by rondiplomatico** [comment]
my experience already shows that even the couple of instructions still needed to install opencmiss are not completely read :-| so yes, i'd also include that as well but would also vote to have this little detection in place; it's easy to check the definition of the OPENCMISS_ROOT env vars etc

**Comment by chrispbradley** [comment]
Yes, if everybody followed instructions there wouldn't be the need for many a thing. If I understand the comment above are you planning on checking to see if the OPENCMISS_ROOT etc. environment variables are set and if so assume an old setup? I'm not sure that would work as we are still going to need these environment variables in the new setup to set binary and library search paths etc.

**Comment by rondiplomatico** [comment]
where in the new setup do we need those? no environment variables required yet, only a convenience "OPENCMISS_INSTALL_DIR" that can be used by the examples to automatically locate an existing installation without having to pass that dir on the command line.

**Comment by chrispbradley** [comment]
The build system only creates a consistent set of binaries and libraries and installs them in a arch path directory under the install dir environment variable. This will not help distinguish, say, what binary to execute (or possibly shared library to load) under multi-config builds. For example, if I have multiple MPI configs installed on my machine and I want to execute an MPI job then I will need to use the correct mpiexec/mpirun that that executable was built with. If I just type mpiexec then it will just run the first mpiexec that it finds in my path which could well be the wrong one. The way unix works is that in order to correctly use the correct binary you need to ensure that the path is set accordingly i.e., there still needs to be some sort of env variables/login script that can do this.

**Comment by rondiplomatico** [comment]
While i see the point for the path of the correct mpiexec, both the mpiexec and example binaries have their "RPATH" set (by cmake install step) to the installed locations of iron etc, so they will automatically load the correct libraries they have been built against. and for mpiexec, (at least on linux/max) the cmake build could automatically create a symlink to the mpi(-exec) the example was built with.

for the planned downloadable pre-compiled binaries i haven't spend any thoughts yet - i dont suppose we also ship out pre-compiled mpi (on linux/mac?), and if not, i wonder how compatiblity with local installations is ensured.

**Comment by chrispbradley** [comment]
Yes, I know the rpaths are set but rpath itself is not completely infaliable. There are some (instrumentation?) libraries that use ldopen calls that will fall back to the LD_LIBRARY_PATH. In addition, and as you mention, there are going to be issues with pre-compiled binaries on some systems with rpath. I don't think the symlink in the example will work as the mpi executables are often scripts that invoke other mpi executables/scripts. There is still a need for the proper unix paths to be set via env vars and a login script (which also does other jobs like setting up the compilers/tools etc.). There also needs to be an easy way of switching configs.

**Comment by hsorby** [comment]
I really think the setting of paths via environment variables and a login script is well outside the manage scripts responsibility, and thus not part of this issue.

The current implementation of the build does not require environment variables so a simple detection of the OPENCMISS_ROOT environment variable would help in detecting another installation of a possibly incompatible opencmiss installation.

In any case this shouldn't be a long term problem as there really should be only one build on a system which is the CMake build.

**Comment by chrispbradley** [comment]
Yes, the setting of variables and login scripts is outside the manage scripts responsibility (with the exception that it may have to act on what the vars are set to). The current build does not require variables however variables and a login script are required to run and use executables. In addition things like setting up, say, the intel compiler or other tools is required in the login script and thus the build system does rely on it. Thus, it may not be possible to simply detect OPENCMISS_ROOT and say it is an old build system. This has yet to be worked out. Agreed that the old build system shouldn't really exist and thus this is not a big problem.
